src/tsHelpers.ts

```diff
@@ -100,3 +100,27 @@ type UnionToIntersection<U> = (U extends any
 export type NoInfer<T> = [T][T extends any ? 0 : never]
 
 export type Omit<T, K extends keyof any> = Pick<T, Exclude<keyof T, K>>
+
+export interface HasMatchFunction<T> {
+  match(v: any): v is T
+}
+
+export const hasMatchFunction = <T>(
+  v: Matcher<T>
+): v is HasMatchFunction<T> => {
+  try {
+    return typeof (v as HasMatchFunction<T>).match === 'function'
+  } catch {
+    return false
+  }
```

Comment:
Yeah, optional chaining is still not an option, since we have tests running back to pre-optional-chaining TS versions.
But I'd say something like 
```suggestion
    return v && typeof (v as HasMatchFunction<T>).match === 'function'
```
would totally suffice here. 
No need to provoke & catch an exception if it can be prevented in the first place.